plugin.audio.mixcloud/lib/history.py

```diff
@@ -0,0 +1,127 @@
+
+Copyright (C) 2011-2020 jackyNIX
+
+This file is part of KODI Mixcloud Plugin.
+
+KODI Mixcloud Plugin is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+KODI Mixcloud Plugin is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with KODI Mixcloud Plugin.  If not, see <http://www.gnu.org/licenses/>.
+'''
+
+
+
+import os
+import sys
+import json
+from datetime import datetime
+import xbmc
+import xbmcaddon
+from .utils import Utils 
+
+
+
+# variables
+__addon__ = xbmcaddon.Addon('plugin.audio.mixcloud')
+CACHED_HISTORY = {}
+
+
+
+class History:
+
+    def __init__(self, name):
+        self.name = name
+        self.data = []
+        self.readFile()
+
+    def readFile(self):
+        starttime = datetime.now()
+        self.data = []
+        filepath = xbmc.translatePath(__addon__.getAddonInfo('profile')) + self.name + '.json'
+        Utils.log('reading json file: ' + filepath)
+        try:
+            # read file
+            if os.path.exists(filepath):
+                text_file = open(filepath, 'r')
+                self.data = json.loads(text_file.read())
+                text_file.close()
+                self.trim()
+            elif __addon__.getSetting(self.name+'_list'):
+                # convert old 2.4.x settings
+                list_data = __addon__.getSetting(self.name + '_list').split(', ')
+                for list_entry in list_data:
+                    json_entry = {}
+                    list_fields = list_entry.split('=')
+                    for list_field in list_fields:
+                        if len(json_entry) == 0:
+                            json_entry['key'] = list_field
+                        elif len(json_entry) == 1:
+                            json_entry['value'] = list_field
+                    self.data.append(json_entry)
+                    self.trim()
+                Utils.log('convert old 2.4.x settings: ' + self.name + ' -> ' + json.dumps(self.data))
+                self.writeFile()
+                __addon__.setSetting(self.name + '_list', None)
+
+        except Exception as e:
+            Utils.log('unable to read json file: ' + filepath, e)
+        elapsedtime = datetime.now() - starttime
+        Utils.log('read ' + str(len(self.data)) + ' items in ' + str(elapsedtime.seconds) + '.' + str(elapsedtime.microseconds) + ' seconds')
+        return self.data
+        
+    def writeFile(self):
+        filepath = xbmc.translatePath(__addon__.getAddonInfo('profile')) + self.name + '.json'
+        try:
+            text_file = open(filepath, 'w+')

```

Comment:
A small remark here. Better do this:
```python
with open(filepath, 'w+') as text_file :
    text_file.write(json.dumps(self.data, indent = 4 * ' '))
```
No closed need then. Now if an exception occurs, the file is not closed.